docs/build/dapp-staking/dapp_staking_architecture.md

```diff
@@ -0,0 +1,672 @@
+
+Unlike v2, there won’t be an option for compounding rewards, at least not directly via an extrinsic call to make compounding rewards work, frontend will need to correctly estimate the reward from claims, and call `lock & stake` extrinsic calls in a batch. The main reason why there’s no intrinsic compounding reward support is that reward claiming isn’t done for a specific contract, but for all of them at once. What dApp would user re-stake their funds on?
+
+The reward for user is calculated using a simple formula:
+
+$user\_reward = total\_stake\_reward * \frac{user\_stake}{total\_staked}$
+
+Rewards are ONLY issued when they are being claimed.
+
+### Claiming Bonus Rewards
+
+If staker staked X **amount** during `voting subperiod` and hasn’t reduced their stake amount below X during `build&earn subperiod`, they will be eligible for extra rewards
+
+$bonus\_reward = total\_bonus\_reward * \frac{user\_contract\_vp\_stake}{total\_vp\_stake}$
+
+The formula demonstrates that bonus reward is calculated per contract and if user was a *loyal staker* for two distinct contracts, they will need to execute two `claim` calls, one for each contract. **vp** refers to `voting subperiod`.
+
+It will only be possible to claim bonus reward for **1 past period**, and only if this period isn’t beyond history depth limit. This shouldn't be a problem, since it's safe to assume that the periods will last longer than a month.
+
+Rewards are ONLY issued when they are being claimed..
+
+### Claiming Dapp Rewards
+
+After an era ends, dApps will be assigned into tiers.
+
+The call footprint is kept the same as in v2, where both smart contract & era need to be specified when calling `claim`.  
+Rewards are ONLY issued when they are being claimed.
+
+## II. Public Calls & Internals
+
+These are the name of the calls exposed by the pallet. Whatever naming is used in frontend can be entirely different.
+
+### 1. Internals
+
+This covers important internal calls which are critical part of the dApps staking v3 design.
+
+#### >> on_initialize
+Executed at the beginning of each block. The solution can be easily extended to utilize scheduler pallet later instead of `on_initialize`.
+
+Checks if new era or new period needs to begin:
+- In case no new era/periods should start from this block, do nothing;
+- If pallet is in maintenance mode, do nothing;
+- In case changes are needed:
+    - **New era during build&earn period:**
+        - calculate dApp tier allocation & reward pool;
+     - assign staker reward pool;
+    - **Final era of build&earn period:**
+        - switch over to the voting period;
+        - this also means switching over to the next period;
+        - calculate new tier configuration;
+    - **New era during voting period:**
+        - switch over to the build&earn period;
+
+:::note
+
+Storage cleanup should occur after switching over to a new period, or can be done lazily via hooks.
+
+:::
```

Comment:
```suggestion
####  on_initialize
Executed at the beginning of each block.

Checks if new era or new period needs to begin:
- In case no new era/periods should start from this block, do nothing;
- If pallet is in maintenance mode, do nothing;
- In case changes are needed:
    - **New era during build&earn period:**
        - calculate dApp tier allocation & reward pool;
        - assign staker reward pool;
    - **Final era of build&earn period:**
        - switch over to the voting period;
        - this also means switching over to the next period;
        - calculate new tier configuration;
    - **New era during voting period:**
        - switch over to the build&earn period;
```